evm/vm/block_info.py

```diff
@@ -0,0 +1,62 @@
+class BlockInfo():
+    # For shard chain, refer to the shard coinbase.
+    _coinbase = None
+
+    # For shard chian, block info of period_start_prevhash.
+    _timestamp = None
+    _number = None
+    _difficulty = None
+    _gas_limit = None
+    _prev_headers = None
```

Comment:
Given all of these are written as direct passthrough `property` methods with no setters, I think it might be better to remove the underscore prefix and property methods and expect them to be accessed directly.

Reply:
I think the purpose of this strategy *is* making these fields won't be changed after initialization. In this way, it will raise `AttributeError` when someone tries to modify these properties. But maybe there's a better approach for this purpose? I'd love to hear your thoughts.

Reply:
I don't think that these will throw errors if modified.  I believe python will happily let you assign over top of `@property` methods.

- The pragmatic part of me says to just use direct properties and we can have *soft* enforcement of not overwriting values (as in, in code reviews and things, we have a pretty strong *culture* of immutability and well defined boundaries when we require mutability).
- The purist likes this idea and suggests making our own `read_only_property` decorator which includes a `setter` which raises such an attribute error.

I'm leaner towards *soft* enforcement, but I'm willing to hear any arguments you have for the later.